set.go

```diff
@@ -0,0 +1,409 @@
+package crdt
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"strings"
+
+	pb "github.com/ipfs/go-ds-crdt/pb"
+
+	ds "github.com/ipfs/go-datastore"
+	query "github.com/ipfs/go-datastore/query"
+)
+
+var (
+	elemsNs        = "s" // elements namespace
+	tombsNs        = "t" // tombstones namespace
+	valueSuffix    = "v"
+	prioritySuffix = "p"
+)
+
+// set implements an Add-Wins Observed-Remove Set using delta-CRDTs
+// (https://arxiv.org/abs/1410.2803) and backing all the data in a
+// go-datastore. It is fully agnostic to MerkleCRDTs or the delta distribution
+// layer.  It chooses the Value with most priority for a Key as the current
+// Value. When two values have the same priority, it chooses by alphabetically
+// sorting their unique IDs alphabetically.
+type set struct {
+	store     ds.Datastore
+	namespace ds.Key
+}
+
+func newCRDTSet(d ds.Datastore, namespace ds.Key) *set {
+	s := &set{
+		namespace: namespace,
+		store:     d,
+	}
+	return s
+}
+
+// Add returns a new delta-set adding the given elements.
+func (s *set) Add(key string, value []byte) *pb.Delta {
+	return &pb.Delta{
+		Elements: []*pb.Element{
+			&pb.Element{
+				Key:   key,
+				Value: value,
+			},
+		},
+		Tombstones: nil,
+	}
+}
+
+// Rmv returns a new delta-set removing the given elements.
+func (s *set) Rmv(key string) (*pb.Delta, error) {
+	delta := &pb.Delta{}
+
+	// /namespace/<key>/elems
+	prefix := s.elemsPrefix(key)
+	q := query.Query{
+		Prefix:   prefix.String(),
+		KeysOnly: true,
+	}
+
+	results, err := s.store.Query(q)
+	if err != nil {
+		return nil, err
+	}
+	defer results.Close()
+
+	for r := range results.Next() {
+		if r.Error != nil {
+			return delta, err
+		}
+
+		id := strings.TrimPrefix(r.Key, prefix.String())
+		delta.Tombstones = append(delta.Tombstones, &pb.Element{
+			Key: key,
+			Id:  id,
+		})
+	}
+	return delta, nil
+}
+
+// Element retrieves the value of an element from the CRDT set.
+func (s *set) Element(key string) ([]byte, error) {
+	// We can only GET an element if it's part of the Set (in
+	// "elems" and not in "tombstones").
+
+	// As an optimization:
+	// * If the element has a value in the store it means:
+	//   -> It occurs at least once in "elems"
+	//   -> It may or not be tombstoned
+	// * If the element does not have a value in the store:
```

Comment:
```suggestion
	// * If the key does not have a value in the store:
```
the comment can be modified accordingly

Reply:
ok